aips/aip-x.md

```diff
@@ -0,0 +1,247 @@
+
+## Alternative Solutions
+
+
+## Specification and Implementation Details
+
+We see intent as a programmable lock for on chain assets. Meaning:
+- One party can declare:
+  - Type of Asset they would like to trade in
+  - Condition when this asset could be released
+- The framework provides mechanism to release the lock when such condition is met.
+
+Logic here can then be translated into a Move struct that looks like the following:
+```
+struct TradeIntent<Source, Args> has key {
+    dispensable_resource: Source,
+    argument: Args,
+    self_delete_ref: DeleteRef,
+    expiry_time: u64,
+    witness_type: TypeInfo,
+}
+```
+
+The `TradeIntent` is parameterized by two type parameters:
+1. `Source`: Value that intent issuer is willing to give away
+2. `Argument`: Intent user will use `Argument` as auxillary info to determine whether the unlock condition is met.
+
+In the `TradeIntent`, we also register for a witness type where the releasing `Source` can only be done when such witness is provided.
+
+The lifecycle of an intent will look like the following:
+1. User sends a transaction to the blockchain to create the trading intent by specifying the resource they can offer and the expected condition when such resource can be unlocked.
+2. The intent would be broadcasted using an event.
+3. Anyone can execute this intent as long as the `Witness` can be provided.
+    - We will expect intent solvers to actively monitor the available intents on chain and create transactions to claim the intent.
+
+With those types being generic, we can implement different trading intents easily:
+- If `Target` is instantiated with `FungibleAsset`, it means we are trading a Move value for a specific fungible asset.
+- If `Target` is instantiated with `LinearTransferRef`, it means we are trading a Move value for an Object.
+- If `Source` is instantiated with `TransferRef`, it means we are giving out ownership to an Object given certain condition.
+- If `Source` is instantiated with `FungibleAsset`, it means we are giving out a certain fungible asset given certain condition.
+   - This doesn't quite work because `FungibleAsset` is not storable. The real implementation would look like giving out ownership of a `FungibleStore`
+
+### Creating intents
+```
+public fun create_intent<Source: store, Args: store + drop, Witness: drop>(
+    dispensable_resource: Source,
+    argument: Args,
+    expiry_time: u64,
+    issuer: address,
+    _witness: Witness,
+): Object<TradeIntent<Source, Args>> 
+```
+User can use this api to register a trading intent to offer a value of `Source` type. The intent can only be executed if anyone can provide a value of `Witness` type. Intent always have an expiry time so that issuer can claim the resource back after the expiry time. Note that the return value of this function is wrapped in an `Object`. This is because we want anyone to be able to claim this intent by providing the address of this object.
```

Comment:
> Intent always have an expiry time so that issuer can claim the resource back after the expiry time.

Isn't it the opposite? issuer should always be able to claim the resource back, but after expiry - matching will be prevented?